### PR TITLE
[ERP-877] Improve UX around Session Timeouts on clinical data forms

### PR DESCRIPTION
**Pull request by ppettitau**
Description of changes:
* Add modal
* Add countdown component to modal that notifies the user approx how many seconds are left in their session
* redirect to the logout screen after it is deemed that a user's session has timed out.

To try it out, add the following to `.env_local` to shorten the session time to 3 minutes, with 1 session refresh after 2 minutes (`SESSION_REFRESH_LEAD_TIME`), and notification at 2.5 mins before final session expiry (`SESSION_EXPIRY_WARNING_LEAD_TIME`).

```
SESSION_COOKIE_AGE = 180
SESSION_REFRESH_MAX_RETRIES = 1
SESSION_REFRESH_LEAD_TIME = 120
SESSION_EXPIRY_WARNING_LEAD_TIME = 150
```

### CODE REVIEW

**Comment by ppettitau** [comment]
To try it out, add the following to `.env_local` to shorten the session time to 3 minutes, with 1 session refresh, and notification at 2 mins before final session expiry.

```
SESSION_COOKIE_AGE = 180
SESSION_REFRESH_MAX_RETRIES = 1
SESSION_REFRESH_LEAD_TIME = 120
```

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 1]
```diff
@@ -540,24 +540,191 @@
 
  </div>
 
-
+    <div id="sessionTimeoutModal" class="modal" tabindex="-1">
+        <div class="modal-dialog">
+            <div class="modal-content">
+                <div class="modal-body">
+                    <p data-session-state="expiring">{% trans "Your session will expire in " %}<strong id="sessionExpiryTimeout">X seconds</strong>.</p>
+                    <p data-session-state="expired" class="d-none alert alert-danger">Your session has expired.</p>
```
Comment (ScriptSmith) [inline rdrf/rdrf/templates/rdrf_cdes/form.html]: Missing translation

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 2]
```diff
@@ -540,24 +540,191 @@
...
+        }
+
+        function clearCountdownTimer() {
+            if (countdownInterval) clearInterval(countdownInterval);
+        }
+
+        function show(secondsLeftInSession) {
+            let secondsUntilNextCountdownUpdate = 30;
+
+                function countdownTimer() {
```
Comment (ScriptSmith) [inline rdrf/rdrf/templates/rdrf_cdes/form.html]: Extra indentation

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 3]
```diff
@@ -540,24 +540,191 @@
...
+            const ONE_MINUTE_IN_SECONDS = 60;
+            let unit = "seconds";
+            let time = secondsLeft;
+
+            if (secondsLeft >= ONE_MINUTE_IN_SECONDS) {
+                time = Math.round(secondsLeft / ONE_MINUTE_IN_SECONDS);
+                unit = time > 1 ? "minutes" : "minute";
+            }
+
+            $sessionTimeoutModal.find("#sessionExpiryTimeout").text(`${time} ${gettext(unit)}`);
```
Comment (ScriptSmith) [inline rdrf/rdrf/templates/rdrf_cdes/form.html]: I think this should technically be done with https://docs.djangoproject.com/en/3.2/topics/i18n/translation/#interpolate for languages that reverse the number and unit, or have different  pluralization style. But let's see how challenging that would be.

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 4]
```diff
@@ -540,24 +540,191 @@
...
+            switch (sessionState) {
+                case SessionState.Expiring: {
+                    $sessionExpiredComponents.addClass("d-none");
+                    $sessionExpiringComponents.removeClass("d-none");
+                    break;
+                }
+                case SessionState.Expired: {
+                    $sessionExpiredComponents.removeClass("d-none");
+                    $sessionExpiringComponents.addClass("d-none");
+                    console.log('showing expired.');
```
Comment (ScriptSmith) [inline rdrf/rdrf/templates/rdrf_cdes/form.html]: Can remove

File: rdrf/rdrf/templatetags/session_info.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from django import template
+from django.conf import settings
+
+register = template.Library()
+
+
+@register.simple_tag()
+def session_info():
```
Comment (ScriptSmith) [inline rdrf/rdrf/templatetags/session_info.py]: Could be in a context processor

https://github.com/eresearchqut/trrf/blob/2c6babdb8515c83d8de329dfbffed904a322ff6d/rdrf/rdrf/context_processors/context_processors.py#L1-L19

**Review by ScriptSmith — CHANGES_REQUESTED** [review]

File: rdrf/rdrf/templates/rdrf_cdes/form.html (hunk 1 above)
Comment (ppettitau) [reply rdrf/rdrf/templates/rdrf_cdes/form.html]: Fixed.

File: rdrf/rdrf/templates/rdrf_cdes/form.html (hunk 2 above)
Comment (ppettitau) [reply rdrf/rdrf/templates/rdrf_cdes/form.html]: Fixed.

File: rdrf/rdrf/templates/rdrf_cdes/form.html (hunk 3 above)
Comment (ppettitau) [reply rdrf/rdrf/templates/rdrf_cdes/form.html]: Good point. Fixed.

File: rdrf/rdrf/templates/rdrf_cdes/form.html (hunk 4 above)
Comment (ppettitau) [reply rdrf/rdrf/templates/rdrf_cdes/form.html]: Fixed.

File: rdrf/rdrf/templatetags/session_info.py (hunk 5 above)
Comment (ppettitau) [reply rdrf/rdrf/templatetags/session_info.py]: Fixed.

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 6]
```diff
@@ -540,24 +539,188 @@
...
+        const sessionTimeoutModal = new bootstrap.Modal($sessionTimeoutModal, {backdrop: "static"});
+        let countdownInterval;
+
+        function updateSessionTimeoutMessage(secondsLeft) {
+            const ONE_MINUTE_IN_SECONDS = 60;
+            let timeoutText = "";
+
+            if (secondsLeft >= ONE_MINUTE_IN_SECONDS) {
+                const minutesLeft = Math.round(secondsLeft / ONE_MINUTE_IN_SECONDS);
+                timeoutText = interpolate(ngettext("%s minute", "%s minutes", minutesLeft), [minutesLeft]);
```
Comment (ScriptSmith) [inline rdrf/rdrf/templates/rdrf_cdes/form.html]: It appears the string is not picked up in the`makemessages` part of the  `build_translation_sources`  script. The JS might need to be partially moved to a static file.

Comment (ppettitau) [reply rdrf/rdrf/templates/rdrf_cdes/form.html]: Done.
```
#: rdrf/rdrf/static/js/session_timeout_manager.js:25
#, javascript-format
msgid "%s minute"
msgid_plural "%s minutes"
msgstr[0] ""
msgstr[1] ""

#: rdrf/rdrf/static/js/session_timeout_manager.js:27
#, javascript-format
msgid "%s second"
msgid_plural "%s seconds"
msgstr[0] ""
msgstr[1] ""
```

**Review by ScriptSmith — APPROVED** [review]

**Review by Lexiel46 — APPROVED** [review]

**Pull request merged**